File: gcode_viewer_v2/scene/view_cube.py
"""Interactive view-cube widget — hover highlights a face, click snaps the
main camera to that face's orthographic view.

Replaces the previous static vtkAnnotatedCubeActor + vtkOrientationMarkerWidget
combo. Building the cube as 6 distinct quad actors lets us pick which face
is under the cursor (vtkPropPicker returns a specific actor) and re-color
it independently. Labels are vtkVectorText oriented per-face so they read
upright when the user looks at that face head-on.

Layout: dedicated renderer in the top-right corner viewport, on its own
layer above the main scene + highlight overlay. The cube's camera mirrors
the main camera's direction-of-view so the cube re-orients in lock-step.
"""

import math
import vtk


# Each main face: (outward_normal, view_up_when_looking_at_face_from_outside)
FACES = {
    "TOP":    ((0.0,  0.0,  1.0), (0.0,  1.0,  0.0)),
    "BOTTOM": ((0.0,  0.0, -1.0), (0.0,  1.0,  0.0)),
    "FRONT":  ((0.0, -1.0,  0.0), (0.0,  0.0,  1.0)),
    "BACK":   ((0.0,  1.0,  0.0), (0.0,  0.0,  1.0)),
    "RIGHT":  ((1.0,  0.0,  0.0), (0.0,  0.0,  1.0)),
    "LEFT":   ((-1.0, 0.0,  0.0), (0.0,  0.0,  1.0)),
}

# 8 corners — chamfered into small triangular click-targets that snap to iso
# views. Each corner is identified by the sign-vector of its position.
# The view_up is +Z for every corner so all iso views show TOP at top.
CORNER_VIEW_UP = (0.0, 0.0, 1.0)
CORNERS = [(sx, sy, sz) for sx in (+1, -1) for sy in (+1, -1) for sz in (+1, -1)]


def _corner_name(sx, sy, sz):
    """Stable key for the 8 corner faces (used for actor lookup + snap dispatch)."""
    return f"CORNER_{sx:+d}_{sy:+d}_{sz:+d}"


CUBE_BODY_COLOR = (0.32, 0.36, 0.42)  # cool gray (matches the previous look)
HIGHLIGHT_COLOR = (0.55, 0.70, 0.95)  # soft blue when hovered
TEXT_COLOR      = (0.95, 0.96, 1.00)
EDGE_COLOR      = (0.18, 0.20, 0.24)  # darker — only used to outline polygons
TEXT_HEIGHT     = 0.30  # text height as fraction of face dimension
CHAMFER_DEPTH   = 0.15  # how much of each half-edge to shave at corners (0..0.5)


# ── Tiny vector helpers (avoids pulling numpy into this module) ─────────────

def _norm(v):
    l = math.sqrt(v[0] ** 2 + v[1] ** 2 + v[2] ** 2)
    return (v[0] / l, v[1] / l, v[2] / l) if l > 1e-12 else (0.0, 0.0, 0.0)


def _cross(a, b):
    return (
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    )


def _add(a, b):
    return (a[0] + b[0], a[1] + b[1], a[2] + b[2])


def _sub(a, b):
    return (a[0] - b[0], a[1] - b[1], a[2] - b[2])


def _scale(v, s):
    return (v[0] * s, v[1] * s, v[2] * s)


def _make_polygon_actor(verts):
    """Build a single-cell polygon actor (octagon, triangle, etc.).

    Edges are drawn via the actor's edge-visibility property — saves us from
    maintaining a separate vtkExtractEdges actor that would have to track
    the chamfered geometry. Flat shading on each face makes the chamfer
    transitions read clearly under normal lighting.
    """
    points = vtk.vtkPoints()
    for v in verts:
        points.InsertNextPoint(*v)

    polygon = vtk.vtkPolygon()
    polygon.GetPointIds().SetNumberOfIds(len(verts))
    for i in range(len(verts)):
        polygon.GetPointIds().SetId(i, i)

    cells = vtk.vtkCellArray()
    cells.InsertNextCell(polygon)

    pd = vtk.vtkPolyData()
    pd.SetPoints(points)
    pd.SetPolys(cells)

    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputData(pd)

    actor = vtk.vtkActor()
    actor.SetMapper(mapper)
    p = actor.GetProperty()
    p.SetColor(*CUBE_BODY_COLOR)
    p.SetInterpolationToFlat()  # show chamfer facets distinctly
    p.SetAmbient(0.45)
    p.SetDiffuse(0.55)
    p.SetSpecular(0.0)
    p.SetEdgeVisibility(True)
    p.SetEdgeColor(*EDGE_COLOR)
    p.SetLineWidth(1.0)
    return actor


# ── ViewCube ────────────────────────────────────────────────────────────────

class ViewCube:
    """Interactive cube widget pinned to the top-right corner of a render window.

    Construct, then call enable() to hook mouse events. Disposal is implicit
    when the render window is destroyed; if you need to detach explicitly,
    call disable().
    """

    def __init__(self, render_window, main_renderer, interactor):
        self.render_window = render_window
        self.main_renderer = main_renderer
        self.main_camera = main_renderer.GetActiveCamera()
        self.interactor = interactor

        # Make sure there's a layer for us above main + overlay
        if render_window.GetNumberOfLayers() < 3:
            render_window.SetNumberOfLayers(3)

        # Dedicated renderer for the cube in the top-right corner
        self.cube_renderer = vtk.vtkRenderer()
        self.cube_renderer.SetLayer(2)
        self.cube_renderer.SetViewport(0.80, 0.78, 1.00, 0.99)
        self.cube_renderer.SetBackgroundAlpha(0.0)
        self.cube_renderer.SetInteractive(False)
        self.cube_renderer.GetActiveCamera().ParallelProjectionOn()  # ortho — looks more stable
        render_window.AddRenderer(self.cube_renderer)

        # Build the cube: 6 face actors, 6 text labels, 1 edges actor
        self._face_actors = {}
        self._build_cube()

        # Sync cube camera to main camera
        self._sync_camera()
        self._cam_observer = self.main_camera.AddObserver(
            "ModifiedEvent", lambda obj, evt: self._sync_camera()
        )

        # Mouse state
        self._hovered = None
        self._mouse_observer = None
        self._click_observer = None
        self._picker = vtk.vtkPropPicker()

    # ── Public API ──────────────────────────────────────────────────────────

    def enable(self):
        if self._mouse_observer is None:
            self._mouse_observer = self.interactor.AddObserver(
                "MouseMoveEvent", self._on_mouse_move
            )
        if self._click_observer is None:
            self._click_observer = self.interactor.AddObserver(
                "LeftButtonPressEvent", self._on_click
            )

    def disable(self):
        if self._mouse_observer is not None:
            self.interactor.RemoveObserver(self._mouse_observer)
            self._mouse_observer = None
        if self._click_observer is not None:
            self.interactor.RemoveObserver(self._click_observer)
            self._click_observer = None

    # ── Cube construction ──────────────────────────────────────────────────

    def _build_cube(self):
        # 6 main faces — each is an octagon (the cube's chamfered cube faces
        # lose 4 corners and gain 4 chamfer edges, becoming 8-sided)
        for name, (normal, view_up) in FACES.items():
            face = self._build_octagon_face(normal, view_up)
            face._face_name = name
            self.cube_renderer.AddActor(face)
            self._face_actors[name] = face

            label = self._build_label(name, normal, view_up)
            label.PickableOff()
            self.cube_renderer.AddActor(label)

        # 8 corner faces — small triangles, one per chamfered corner
        for (sx, sy, sz) in CORNERS:
            cname = _corner_name(sx, sy, sz)
            corner_actor = self._build_corner_triangle(sx, sy, sz)
            corner_actor._face_name = cname
            self.cube_renderer.AddActor(corner_actor)
            self._face_actors[cname] = corner_actor

    def _build_octagon_face(self, normal, view_up):
        """A single octagonal face — a cube face with its 4 corners chamfered.

        Vertices are emitted in CCW order (viewed from +n) so the polygon's
        outward normal is correct without manual normal management.
        """
        n = _norm(normal)
        vu = _norm(view_up)
        right = _cross(vu, n)

        center = _scale(n, 0.5)
        h = 0.5
        d = CHAMFER_DEPTH

        # 8 octagon vertices (u, v) coefficients in CCW order from +n
        uv = [
            ( h - d,  h),
            (-(h - d),  h),
            (-h,  h - d),
            (-h, -(h - d)),
            (-(h - d), -h),
            ( h - d, -h),
            ( h, -(h - d)),
            ( h,  h - d),
        ]
        verts = [
            _add(center, _add(_scale(right, u), _scale(vu, v)))
            for (u, v) in uv
        ]
        return _make_polygon_actor(verts)

    def _build_corner_triangle(self, sx, sy, sz):
        """One of the 8 chamfered-corner triangles. Outward normal is the
        corner's diagonal direction (sx, sy, sz)/√3.
        """
        h = 0.5
        d = CHAMFER_DEPTH
        # 3 vertices on the 3 cube edges meeting at this corner, each shifted
        # inward by `d` from the corner.
        v_along_x = (sx * (h - d), sy * h, sz * h)
        v_along_y = (sx * h, sy * (h - d), sz * h)
        v_along_z = (sx * h, sy * h, sz * (h - d))

        # CCW-from-outside winding flips with the parity of the sign product.
        # See cross-product derivation in the module docstring.
        if sx * sy * sz > 0:
            verts = [v_along_x, v_along_y, v_along_z]
        else:
            verts = [v_along_x, v_along_z, v_along_y]
        return _make_polygon_actor(verts)

    def _build_label(self, name, normal, view_up):
        """Text label that lies flat on a face and reads upright when viewed
        from outside the face."""
        text = vtk.vtkVectorText()
        text.SetText(name)
        text.Update()
        bounds = text.GetOutput().GetBounds()
        text_w = bounds[1] - bounds[0]
        text_h = bounds[3] - bounds[2]

        # Scale so the text height fits TEXT_HEIGHT of a face dimension,
        # but never let the width exceed 0.85 of face dimension.
        max_w = 0.85
        if text_h > 0 and text_w > 0:
            scale = min(TEXT_HEIGHT / text_h, max_w / text_w)
        else:
            scale = TEXT_HEIGHT

        n = _norm(normal)
        vu = _norm(view_up)
        right = _cross(vu, n)

        # Rotation matrix mapping vtkVectorText's local axes to the face's
        # axes: text-X → right, text-Y → vu, text-Z → n.
        rot = vtk.vtkMatrix4x4()
        rot.Identity()
        for i in range(3):
            rot.SetElement(i, 0, right[i])
            rot.SetElement(i, 1, vu[i])
            rot.SetElement(i, 2, n[i])

        # Build the transform in PostMultiply mode so calls apply in order.
        # 1) Center the text on its own origin
        # 2) Scale to fit
        # 3) Rotate to face orientation
        # 4) Translate just outside the face surface (avoid z-fighting)
        offset = 0.005
        world_pos = _scale(n, 0.5 + offset)

        transform = vtk.vtkTransform()
        transform.PostMultiply()
        transform.Translate(-text_w / 2.0, -text_h / 2.0, 0.0)
        transform.Scale(scale, scale, scale)
        transform.Concatenate(rot)
        transform.Translate(*world_pos)

        tf = vtk.vtkTransformPolyDataFilter()
        tf.SetInputConnection(text.GetOutputPort())
        tf.SetTransform(transform)
        tf.Update()

        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputConnection(tf.GetOutputPort())

        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        actor.GetProperty().SetColor(*TEXT_COLOR)
        actor.GetProperty().SetLighting(False)
        return actor

    # ── Camera sync ────────────────────────────────────────────────────────

    def _sync_camera(self):
        """Mirror the main camera's view direction onto the cube's camera so
        the cube's orientation reflects what the main scene is showing."""
        cube_cam = self.cube_renderer.GetActiveCamera()
        mp = self.main_camera.GetPosition()
        mf = self.main_camera.GetFocalPoint()
        d = _sub(mp, mf)
        l = math.sqrt(d[0] ** 2 + d[1] ** 2 + d[2] ** 2)
        if l < 1e-9:
            return
        # Place cube cam at unit-distance in the same direction. The cube is
        # ~1 unit across; 4 units of distance gives a comfortable framing.
        n = (d[0] / l, d[1] / l, d[2] / l)
        cube_cam.SetPosition(n[0] * 4.0, n[1] * 4.0, n[2] * 4.0)
        cube_cam.SetFocalPoint(0.0, 0.0, 0.0)
        cube_cam.SetViewUp(*self.main_camera.GetViewUp())
        # Parallel-projection scale: the cube spans 1 unit, half-width is 0.5;
        # at iso, the visible diagonal extends to ~0.87. Use 0.95 to leave a
        # comfortable margin on every edge regardless of camera angle.
        cube_cam.SetParallelScale(0.95)
        self.cube_renderer.ResetCameraClippingRange()

    # ── Mouse handling ─────────────────────────────────────────────────────

    def _is_in_viewport(self, x, y):
        """Is screen position (x, y) inside the cube's corner viewport?"""
        size = self.render_window.GetSize()
        if size[0] == 0 or size[1] == 0:
            return False
        nx = x / size[0]
        ny = y / size[1]
        vp = self.cube_renderer.GetViewport()
        return vp[0] <= nx <= vp[2] and vp[1] <= ny <= vp[3]

    def _pick_face(self, x, y):
        if not self._is_in_viewport(x, y):
            return None
        self._picker.Pick(x, y, 0, self.cube_renderer)
        prop = self._picker.GetActor()
        if prop is None:
            return None
        return getattr(prop, "_face_name", None)

    def _on_mouse_move(self, iren, event):
        x, y = iren.GetEventPosition()
        face = self._pick_face(x, y)
        if face != self._hovered:
            if self._hovered is not None:
                self._face_actors[self._hovered].GetProperty().SetColor(*CUBE_BODY_COLOR)
            if face is not None:
                self._face_actors[face].GetProperty().SetColor(*HIGHLIGHT_COLOR)
            self._hovered = face
            self.render_window.Render()

    def _on_click(self, iren, event):
        x, y = iren.GetEventPosition()
        face = self._pick_face(x, y)
        if face is not None:
            self._snap_to_face(face)

    def _snap_to_face(self, face_name):
        """Move the main camera so it views the named face dead-on (or, for a
        chamfered corner, looks along the corner's diagonal so the three
        adjacent main faces are all visible). Preserves focal point and
        distance — only direction changes."""
        if face_name in FACES:
            normal, view_up = FACES[face_name]
            normal = _norm(normal)
        elif face_name.startswith("CORNER_"):
            # Parse "CORNER_+1_-1_+1" → (1, -1, 1)
            parts = face_name.split("_")
            sx, sy, sz = int(parts[1]), int(parts[2]), int(parts[3])
            normal = _norm((float(sx), float(sy), float(sz)))
            view_up = CORNER_VIEW_UP
        else:
            return  # unknown name — nothing to snap to

        focal = self.main_camera.GetFocalPoint()
        pos = self.main_camera.GetPosition()
        d = _sub(pos, focal)
        dist = math.sqrt(d[0] ** 2 + d[1] ** 2 + d[2] ** 2)
        if dist < 1e-9:
            dist = 100.0
        new_pos = _add(focal, _scale(normal, dist))
        self.main_camera.SetPosition(*new_pos)
        self.main_camera.SetViewUp(*view_up)
        self.main_renderer.ResetCameraClippingRange()
        self.render_window.Render()

File: gcode_viewer_v3/internal/scene/view_cube.go
// Interactive view-cube widget — port of gcode_viewer_v2/scene/view_cube.py.
//
// 14 pickable polygons make up the cube:
//   - 6 octagonal main faces (each cube face has its 4 corners chamfered)
//   - 8 triangular corner faces (one per chamfered corner)
//
// Click a main face → main camera snaps to that orthogonal view.
// Click a corner → main camera snaps to the iso view that shows the three
// adjacent main faces simultaneously.
//
// The cube's camera mirrors the main camera so the cube re-orients in
// lock-step as the user orbits.
//
// Text labels are not yet ported — they require bundling a TTF font through
// g3n's text package and laying out per-face transforms. The chamfered
// geometry (with face shading via per-vertex normals) reads clearly enough
// without labels for now; labels land in a follow-up alongside the M5
// toolbar polish.

package scene

import (
	"fmt"

	"github.com/g3n/engine/camera"
	"github.com/g3n/engine/core"
	"github.com/g3n/engine/experimental/collision"
	"github.com/g3n/engine/geometry"
	"github.com/g3n/engine/gls"
	"github.com/g3n/engine/graphic"
	"github.com/g3n/engine/light"
	"github.com/g3n/engine/material"
	"github.com/g3n/engine/math32"
)

// ChamferDepth is how far in from each cube corner the chamfer extends, in
// units where the cube spans -0.5 to +0.5. Matches the v2 constant.
const ChamferDepth = 0.15

// Color palette — matches v2's view_cube.py.
var (
	cubeBodyColor    = math32.Color{R: 0.32, G: 0.36, B: 0.42} // cool gray
	cubeHighlightCol = math32.Color{R: 0.55, G: 0.70, B: 0.95} // soft blue
	cubeEdgeColor    = math32.Color{R: 0.10, G: 0.11, B: 0.13} // near-black outline
)

// ViewCubeFace is one pickable polygon of the cube — either a main octagonal
// face or a corner triangle. Normal and Up describe how the main camera
// should snap when this face is clicked.
//
// Group bundles the face mesh and its edge outline so we can add a single
// node to the scene per face. UserData on the Mesh (NOT the Group) is what
// the Raycaster recovers via inter.Object.GetNode().UserData().
type ViewCubeFace struct {
	Group  *core.Node
	Mesh   *graphic.Mesh
	Edges  *graphic.LineStrip
	Mat    *material.Standard
	Name   string         // "TOP", "BOTTOM", ..., or "CORNER_+1_-1_+1"
	Normal math32.Vector3 // outward normal in cube-local space
	Up     math32.Vector3 // view-up direction when looking AT this face
}

// ViewCube is the interactive widget rendered into a corner viewport.
type ViewCube struct {
	Scene  *core.Node
	Camera *camera.Camera
	Faces  []*ViewCubeFace

	hovered *ViewCubeFace
}

// faceSpec describes one of the 6 main cube faces.
type faceSpec struct {
	name   string
	normal math32.Vector3
	up     math32.Vector3
}

var mainFaceSpecs = []faceSpec{
	{"TOP", math32.Vector3{X: 0, Y: 0, Z: 1}, math32.Vector3{X: 0, Y: 1, Z: 0}},
	{"BOTTOM", math32.Vector3{X: 0, Y: 0, Z: -1}, math32.Vector3{X: 0, Y: 1, Z: 0}},
	{"FRONT", math32.Vector3{X: 0, Y: -1, Z: 0}, math32.Vector3{X: 0, Y: 0, Z: 1}},
	{"BACK", math32.Vector3{X: 0, Y: 1, Z: 0}, math32.Vector3{X: 0, Y: 0, Z: 1}},
	{"RIGHT", math32.Vector3{X: 1, Y: 0, Z: 0}, math32.Vector3{X: 0, Y: 0, Z: 1}},
	{"LEFT", math32.Vector3{X: -1, Y: 0, Z: 0}, math32.Vector3{X: 0, Y: 0, Z: 1}},
}

// cornerSign holds the ±1 sign of each axis for one of the 8 cube corners.
type cornerSign struct{ Sx, Sy, Sz int }

var cornerSigns = []cornerSign{
	{+1, +1, +1}, {+1, +1, -1}, {+1, -1, +1}, {+1, -1, -1},
	{-1, +1, +1}, {-1, +1, -1}, {-1, -1, +1}, {-1, -1, -1},
}

// cornerViewUp is the view-up direction used when snapping to ANY corner ISO
// view. +Z keeps the world's "up" reading at the top of every iso shot.
var cornerViewUp = math32.Vector3{X: 0, Y: 0, Z: 1}

// NewViewCube builds the chamfered cube widget.
func NewViewCube() *ViewCube {
	vc := &ViewCube{Scene: core.NewNode()}

	// Narrow-FOV perspective so the cube reads close to "orthographic" but
	// the Raycaster's SetFromCamera math (which assumes a single-origin
	// convergent ray) gives correct results. With orthographic the rays
	// should be parallel — g3n's SetFromCamera doesn't model that, which
	// caused the hover to pick the wrong face for off-center cursor
	// positions. Near/far are tight around the cube to keep depth precision
	// excellent at this scale.
	vc.Camera = camera.NewPerspective(1, 0.5, 100, 30, camera.Vertical)
	vc.Scene.Add(vc.Camera)

	// Pure full-strength ambient, no directional light. Combined with
	// material AmbientColor = body color (set in newPolygonMesh), every
	// face renders as a flat uniform body color regardless of orientation.
	// The chamfer transitions stay visible thanks to the edge outlines.
	vc.Scene.Add(light.NewAmbient(&math32.Color{R: 1, G: 1, B: 1}, 1.0))

	for _, spec := range mainFaceSpecs {
		f := buildOctagonFace(spec)
		vc.Faces = append(vc.Faces, f)
		vc.Scene.Add(f.Group)
	}
	for _, c := range cornerSigns {
		f := buildCornerTriangle(c)
		vc.Faces = append(vc.Faces, f)
		vc.Scene.Add(f.Group)
	}

	return vc
}

// buildOctagonFace builds one of the 6 chamfered cube faces (an octagon).
//
// Vertices are emitted in CCW order (viewed from +n), giving the face an
// outward-pointing normal without manual normal management.
func buildOctagonFace(spec faceSpec) *ViewCubeFace {
	n := normalize3(spec.normal)
	vu := normalize3(spec.up)
	right := cross3(vu, n)
	center := math32.Vector3{X: n.X * 0.5, Y: n.Y * 0.5, Z: n.Z * 0.5}

	const h = 0.5
	const d = ChamferDepth

	// 8 octagon vertices in (right, vu) plane coefficients, CCW from +n.
	uv := [8][2]float32{
		{h - d, h},
		{-(h - d), h},
		{-h, h - d},
		{-h, -(h - d)},
		{-(h - d), -h},
		{h - d, -h},
		{h, -(h - d)},
		{h, h - d},
	}
	verts := make([]math32.Vector3, 8)
	for i, p := range uv {
		u, v := p[0], p[1]
		verts[i] = math32.Vector3{
			X: center.X + right.X*u + vu.X*v,
			Y: center.Y + right.Y*u + vu.Y*v,
			Z: center.Z + right.Z*u + vu.Z*v,
		}
	}

	mesh, mat := newPolygonMesh(verts, n, &cubeBodyColor)
	edges := newPolygonEdges(verts, n)

	group := core.NewNode()
	group.Add(mesh)
	group.Add(edges)

	// Text label for the 6 main faces (TOP, BOTTOM, ...). Reads upright
	// when the user looks at the face from outside the cube — vu defines
	// the "up" axis on the face surface.
	if label := newFaceLabel(spec.name, n, vu); label != nil {
		group.Add(label)
	}

	f := &ViewCubeFace{
		Group:  group,
		Mesh:   mesh,
		Edges:  edges,
		Mat:    mat,
		Name:   spec.name,
		Normal: n,
		Up:     vu,
	}
	mesh.SetUserData(f)
	return f
}

// buildCornerTriangle builds the small triangle at one chamfered corner of
// the cube. Outward normal is the corner's diagonal (sx, sy, sz)/√3.
func buildCornerTriangle(c cornerSign) *ViewCubeFace {
	const h = 0.5
	const d = ChamferDepth

	sx, sy, sz := float32(c.Sx), float32(c.Sy), float32(c.Sz)

	vAlongX := math32.Vector3{X: sx * (h - d), Y: sy * h, Z: sz * h}
	vAlongY := math32.Vector3{X: sx * h, Y: sy * (h - d), Z: sz * h}
	vAlongZ := math32.Vector3{X: sx * h, Y: sy * h, Z: sz * (h - d)}

	// CCW-from-outside winding flips with the parity of the sign product.
	// (Same derivation as Python view_cube.py.)
	var verts []math32.Vector3
	if c.Sx*c.Sy*c.Sz > 0 {
		verts = []math32.Vector3{vAlongX, vAlongY, vAlongZ}
	} else {
		verts = []math32.Vector3{vAlongX, vAlongZ, vAlongY}
	}

	normal := normalize3(math32.Vector3{X: sx, Y: sy, Z: sz})

	mesh, mat := newPolygonMesh(verts, normal, &cubeBodyColor)
	edges := newPolygonEdges(verts, normal)

	group := core.NewNode()
	group.Add(mesh)
	group.Add(edges)

	name := fmt.Sprintf("CORNER_%+d_%+d_%+d", c.Sx, c.Sy, c.Sz)
	f := &ViewCubeFace{
		Group:  group,
		Mesh:   mesh,
		Edges:  edges,
		Mat:    mat,
		Name:   name,
		Normal: normal,
		Up:     cornerViewUp,
	}
	mesh.SetUserData(f)
	return f
}

// newPolygonMesh builds a g3n Mesh from a CCW-ordered vertex ring with a
// uniform per-vertex normal (so g3n's lighting shows the face as flat-shaded
// without averaging across adjacent meshes).
//
// Triangulation is a simple fan from vertex 0, which is correct for any
// convex polygon — both octagons and triangles are convex.
func newPolygonMesh(verts []math32.Vector3, normal math32.Vector3, color *math32.Color) (*graphic.Mesh, *material.Standard) {
	geom := geometry.NewGeometry()

	positions := math32.NewArrayF32(0, len(verts)*3)
	normals := math32.NewArrayF32(0, len(verts)*3)
	for _, v := range verts {
		positions.Append(v.X, v.Y, v.Z)
		normals.Append(normal.X, normal.Y, normal.Z)
	}

	indices := math32.NewArrayU32(0, (len(verts)-2)*3)
	for i := 1; i < len(verts)-1; i++ {
		indices.Append(0, uint32(i), uint32(i+1))
	}

	geom.AddVBO(gls.NewVBO(positions).AddAttrib(gls.VertexPosition))
	geom.AddVBO(gls.NewVBO(normals).AddAttrib(gls.VertexNormal))
	geom.SetIndices(indices)

	colorCopy := *color
	mat := material.NewStandard(&colorCopy)
	// Ambient color = body color so the face renders as flat uniform
	// color under the full-strength ambient light (no diffuse shading
	// since the cube scene has no directional light).
	mat.SetAmbientColor(&colorCopy)
	mat.SetSide(material.SideDouble)
	mat.SetShininess(0)

	return graphic.NewMesh(geom, mat), mat
}

// SyncToMainCamera mirrors the main camera's view direction onto the cube
// camera so the cube re-orients in lock-step with the main scene.
//
// `focal` is the point the main camera is currently orbiting (set by the
// main view's last LookAt — tracked by the UI layer in sceneState.focal).
func (vc *ViewCube) SyncToMainCamera(mainCam *camera.Camera, focal math32.Vector3) {
	mp := mainCam.Position()

	rx := mp.X - focal.X
	ry := mp.Y - focal.Y
	rz := mp.Z - focal.Z

	mag := math32.Sqrt(rx*rx + ry*ry + rz*rz)
	if mag < 1e-6 {
		return
	}
	// At distance 3.5 with the 30° vertical FOV the unit cube fills ~50% of
	// the viewport — readable but not overwhelming. Tweak this number with
	// the FOV in NewViewCube if the cube needs to look bigger or smaller.
	scale := float32(3.5) / mag
	vc.Camera.SetPosition(rx*scale, ry*scale, rz*scale)
	vc.Camera.LookAt(
		&math32.Vector3{X: 0, Y: 0, Z: 0},
		&math32.Vector3{X: 0, Y: 0, Z: 1},
	)
}

// SetHovered updates which face shows the highlight color (or nil to clear).
//
// Since the cube has no directional light, only AmbientColor contributes to
// the rendered color — but we update both AmbientColor and the base material
// Color so the face stays consistent if someone later re-introduces a
// directional light.
func (vc *ViewCube) SetHovered(face *ViewCubeFace) {
	if vc.hovered == face {
		return
	}
	if vc.hovered != nil {
		body := cubeBodyColor
		vc.hovered.Mat.SetColor(&body)
		vc.hovered.Mat.SetAmbientColor(&body)
	}
	if face != nil {
		hi := cubeHighlightCol
		face.Mat.SetColor(&hi)
		face.Mat.SetAmbientColor(&hi)
	}
	vc.hovered = face
}

// PickFace casts a ray from the cube camera through the given normalized
// device coordinates (NDC, both in [-1, 1]) and returns the hit face, or
// nil if no face was hit.
func (vc *ViewCube) PickFace(ndcX, ndcY float32) *ViewCubeFace {
	rc := collision.NewRaycaster(
		&math32.Vector3{X: 0, Y: 0, Z: 0},
		&math32.Vector3{X: 0, Y: 0, Z: -1},
	)
	if err := rc.SetFromCamera(vc.Camera, ndcX, ndcY); err != nil {
		return nil
	}

	intersects := rc.IntersectObjects(vc.Scene.Children(), true)
	for _, inter := range intersects {
		if inter.Object == nil {
			continue
		}
		if face, ok := inter.Object.GetNode().UserData().(*ViewCubeFace); ok {
			return face
		}
	}
	return nil
}

// newPolygonEdges builds a closed-loop LineStrip tracing the polygon's
// perimeter. Vertices are nudged outward along the polygon's normal by a
// tiny offset so the edge line draws cleanly in front of the face mesh
// (otherwise z-fighting causes visible flicker).
func newPolygonEdges(verts []math32.Vector3, normal math32.Vector3) *graphic.LineStrip {
	const nudge = 0.002

	n := len(verts)
	positions := math32.NewArrayF32(0, (n+1)*3)
	colors := math32.NewArrayF32(0, (n+1)*3)

	for i := 0; i <= n; i++ {
		v := verts[i%n]
		positions.Append(
			v.X+normal.X*nudge,
			v.Y+normal.Y*nudge,
			v.Z+normal.Z*nudge,
		)
		colors.Append(cubeEdgeColor.R, cubeEdgeColor.G, cubeEdgeColor.B)
	}

	geom := geometry.NewGeometry()
	geom.AddVBO(gls.NewVBO(positions).AddAttrib(gls.VertexPosition))
	geom.AddVBO(gls.NewVBO(colors).AddAttrib(gls.VertexColor))
	return graphic.NewLineStrip(geom, material.NewBasic())
}

// ── small vector helpers (kept local to avoid pulling math elsewhere) ──

func normalize3(v math32.Vector3) math32.Vector3 {
	l := math32.Sqrt(v.X*v.X + v.Y*v.Y + v.Z*v.Z)
	if l < 1e-12 {
		return math32.Vector3{}
	}
	return math32.Vector3{X: v.X / l, Y: v.Y / l, Z: v.Z / l}
}

func cross3(a, b math32.Vector3) math32.Vector3 {
	return math32.Vector3{
		X: a.Y*b.Z - a.Z*b.Y,
		Y: a.Z*b.X - a.X*b.Z,
		Z: a.X*b.Y - a.Y*b.X,
	}
}
